pkg/reconciler/pipeline/pipeline.go

```diff
@@ -0,0 +1,297 @@
+) *controller.Impl {
+
+	r := &Reconciler{
+		Base:                reconciler.NewBase(opt, controllerAgentName),
+		pipelineLister:      pipelineInformer.Lister(),
+		addressableInformer: addressableInformer,
+		subscriptionLister:  subscriptionInformer.Lister(),
+	}
+	impl := controller.NewImpl(r, r.Logger, ReconcilerName)
+
+	r.Logger.Info("Setting up event handlers")
+
+	r.tracker = tracker.New(impl.EnqueueKey, opt.GetTrackerLease())
+	pipelineInformer.Informer().AddEventHandler(controller.HandleAll(impl.Enqueue))
+
+	// Register handler for Subscriptions that are owned by Pipeline, so that
+	// we get notified if they change.
+	subscriptionInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controller.Filter(v1alpha1.SchemeGroupVersion.WithKind("Pipeline")),
+		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
+	})
+
+	return impl
+}
+
+// Reconcile compares the actual state with the desired, and attempts to
+// reconcile the two. It then updates the Status block of the Pipeline resource
+// with the current Status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logging.FromContext(ctx).Error("reconciling", zap.String("key", key))
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
+		return nil
+	}
+
+	// Get the Pipeline resource with this namespace/name
+	original, err := r.pipelineLister.Pipelines(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("Pipeline key in work queue no longer exists")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	pipeline := original.DeepCopy()
+
+	// Reconcile this copy of the Pipeline and then write back any status
+	// updates regardless of whether the reconcile error out.
+	reconcileErr := r.reconcile(ctx, pipeline)
+	if reconcileErr != nil {
+		logging.FromContext(ctx).Error("Error reconciling Pipeline", zap.Error(reconcileErr))
+		r.Recorder.Eventf(pipeline, corev1.EventTypeWarning, reconcileFailed, "Pipeline reconciliation failed: %v", reconcileErr)
+	} else {
+		logging.FromContext(ctx).Debug("Successfully reconciled Pipeline")
+		r.Recorder.Eventf(pipeline, corev1.EventTypeNormal, reconciled, "Pipeline reconciled")
+	}
+
+	if _, updateStatusErr := r.updateStatus(ctx, pipeline.DeepCopy()); updateStatusErr != nil {
+		logging.FromContext(ctx).Warn("Error updating Pipeline status", zap.Error(updateStatusErr))
+		r.Recorder.Eventf(pipeline, corev1.EventTypeWarning, updateStatusFailed, "Failed to update pipeline status: %s", key)
+		return updateStatusErr
+	}
+
+	// Requeue if the resource is not ready:
+	return reconcileErr
+}
+
+func pipelineSubscriptionName(pipelineName string, step int) string {
```

Comment:
Redundant with resources/subscription.go's `PipelineSubscriptionName`. Remove one of them (my preference is to remove this one).

Reply:
yeah, I moved it there and changed refs to it, but forgot to remove this from here, removed.